### PR TITLE
Multiversion support

### LINKED ISSUE #116: Support multiple versions
Currently it seems that `sbt-microsites` is built only from master branch. There should be a mechanism to support multiple versions that can be selected by a dropdown or similar navigation element. Versions could be generated from git tags or from configuration file which maps versions to git branches or commits.

### PR DESCRIPTION
**Pull request by calvellido**
* Set a new `makeMultiversionMicrosite` task.
* Add new `micrositeVersionList` setting.
* `sbt` tasks reorganization in order to make way for the new task, while keeping same behaviour instact.
* Add a bunch of tasks new auxiliary functions.
* Include logic necessary (layout/JS) for when a multiversion selector is included.
* Set proper styling for the version selector. 
* Update docs.

This closes #116

### CODE REVIEW

**Comment by calvellido** [comment]
This is ready to rumble now :rocket:

File: src/main/scala/microsites/MicrositeKeys.scala [hunk 1]
```diff
@@ -167,14 +201,64 @@ trait MicrositeAutoImportSettings extends MicrositeKeys {
...
+      .map(version => {
+        Version(version, own = (ownVersion == version))
+      })
+      .toList
+  }
+
+  def pluginProjects(pluginName: String): Option[Array[String]] = {
+    val sbtPluginsOutput = "sbt --error plugins".lineStream
+    val pluginLine =
+      sbtPluginsOutput.find(line => line.trim.startsWith(s"$pluginName: enabled in "))
```
Comment (juanpedromoreno) [inline src/main/scala/microsites/MicrositeKeys.scala]:
```suggestion
      sbtPluginsOutput.find(_.trim.startsWith(s"$pluginName: enabled in "))
```

File: src/main/scala/microsites/MicrositeKeys.scala [hunk 2]
```diff
@@ -167,14 +201,64 @@ trait MicrositeAutoImportSettings extends MicrositeKeys {
...
+    val jekyllDir  = "jekyll"
+    val targetPath = s"$targetDir$jekyllDir/_data/versions.json"
+    createFile(targetPath)
+    writeContentToFile(content.asJson.toString, targetPath)
+    targetPath.toFile
+  }
+
+  def generateVersionList(versionStringList: List[String], ownVersion: String) = {
+    versionStringList
+      .map(version => {
```
Comment (juanpedromoreno) [inline src/main/scala/microsites/MicrositeKeys.scala]: The braces wouldn't be needed

File: src/main/scala/microsites/MicrositeKeys.scala [hunk 3]
```diff
@@ -167,14 +201,64 @@ trait MicrositeAutoImportSettings extends MicrositeKeys {
 
   lazy val fr = new FileReader
 
+  def createVersionsJson(targetDir: String, content: List[Version]): File = {
+    val jekyllDir  = "jekyll"
+    val targetPath = s"$targetDir$jekyllDir/_data/versions.json"
+    createFile(targetPath)
+    writeContentToFile(content.asJson.toString, targetPath)
```
Comment (juanpedromoreno) [inline src/main/scala/microsites/MicrositeKeys.scala]: You might want to write the json with a circe Printer:

```scala
content.asJson.pretty(Printer.noSpaces) // Or
content.asJson.pretty(Printer.spaces2)
```

File: src/main/scala/microsites/MicrositeKeys.scala [hunk 4]
```diff
@@ -167,14 +201,64 @@ trait MicrositeAutoImportSettings extends MicrositeKeys {
...
+
+  def pluginProjects(pluginName: String): Option[Array[String]] = {
+    val sbtPluginsOutput = "sbt --error plugins".lineStream
+    val pluginLine =
+      sbtPluginsOutput.find(line => line.trim.startsWith(s"$pluginName: enabled in "))
+
+    pluginLine match {
+      case Some(line) => Some(line.trim.stripPrefix(s"$pluginName: enabled in ").split(", "))
+      case None       => None
+    }
```
Comment (juanpedromoreno) [inline src/main/scala/microsites/MicrositeKeys.scala]:
```scala
pluginLine.map(_.trim.stripPrefix(s"$pluginName: enabled in ").split(", "))
```

?

File: src/main/scala/microsites/MicrositeKeys.scala [hunk 5]
```diff
@@ -167,14 +201,64 @@ trait MicrositeAutoImportSettings extends MicrositeKeys {
...
+  }
+
+  // Generate a microsite externally through sbt and sbt-microsites tasks
+  def createMicrositeVersion(
+      sourceDir: String,
+      targetDir: String,
+      baseUrl: String,
+      version: String): Unit = {
+    val newBaseUrl =
+      if (version != "") s"${baseUrl}/$version" else s"${baseUrl}"
```
Comment (juanpedromoreno) [inline src/main/scala/microsites/MicrositeKeys.scala]:
```suggestion
      if (version.nonEmpty) s"$baseUrl/$version" else baseUrl
```

File: src/main/scala/microsites/MicrositeKeys.scala [hunk 6]
```diff
@@ -167,14 +201,64 @@ trait MicrositeAutoImportSettings extends MicrositeKeys {
...
+    val pluginName            = "microsites.MicrositesPlugin"
+    val sbtMicrositesProjects = pluginProjects(pluginName)
+    sbtMicrositesProjects match {
+      case Some(projects) => {
+        List(
+          "sbt",
+          s"""clean; set ${projects(0)}/micrositeBaseUrl := "$newBaseUrl"; ${projects(0)}/makeMicrosite""").!
+        Files.move(
+          FileSystems.getDefault().getPath(sourceDir),
+          FileSystems.getDefault().getPath(s"$targetDir/$version"),
```
Comment (juanpedromoreno) [inline src/main/scala/microsites/MicrositeKeys.scala]: I believe you can use:

```scala
Paths.get(pathString)
```

in both cases above.

File: docs/docs/docs/build-the-microsite.md [hunk 7]
```diff
@@ -12,7 +12,28 @@ Once you have written your documents, you can build the microsite running this s
 sbt> makeMicrosite
 ```
 
-Internally, it'll sequentially run other tasks including its own, [`tut`](https://github.com/tpolecat/tut), and `makeSite` ([sbt-site](https://github.com/sbt/sbt-site)) tasks.
+Internally, it'll sequentially run other tasks including its own, [`tut`](https://github.com/tpolecat/tut) or [`mdoc`](https://scalameta.org/mdoc/), and `makeSite` ([sbt-site](https://github.com/sbt/sbt-site)) tasks.
+
+
+# Multiversion support
+
+Starting on version `1.0.0` you can build microsites including different versions of your library. This is available for the `light` theme, and you need to have [`git`](https://git-scm.com/) installed and available in your system. Most probably you already have it, but in other case please check your system docs, or go to the [git website](https://git-scm.com/download/) on how to proceed.
```
Comment (juanpedromoreno) [inline docs/docs/docs/build-the-microsite.md]: Do you think it could be possible to use this feature in other themes down the road?

Comment (calvellido) [reply docs/docs/docs/build-the-microsite.md]: Definitely!  It's only not being offered to the `pattern` theme at this time, along more features that are not ready yet, but I hope to bring them all over.

This PR was taking longer than I was expecting, so I reduce the scope a bit for this release.

File: src/main/scala/microsites/MicrositeKeys.scala (hunk 1 above)
Comment (calvellido) [reply src/main/scala/microsites/MicrositeKeys.scala]: Ha!  And I was asking myself how could I use that underscore syntax... 😅

File: src/main/scala/microsites/MicrositeKeys.scala (hunk 4 above)
Comment (calvellido) [reply src/main/scala/microsites/MicrositeKeys.scala]: Would the use of `map` there return a `None` in the case that `pluginLine` was one?  I was under the assumption that `map` would only work with _right_ values, but I wasn't sure what would happen for `None` value.

**Comment by calvellido** [comment]
@juanpedromoreno thanks for the review, I'll keep addressing the comments later.

Not specifically related with the review, but do you know if there is any aux function to delete a directory (empty or not) in `sbt-org-policies` (or any other dependency in the project for that matter)?  It could have been of help here (to delete the temporary directory where the microsites are managed until they are moved to their final destination), but I wasn't sure if I should be implementing it here.

**Comment by juanpedromoreno** [comment]
@calvellido Maybe you could:

```scala
new scala.reflect.io.Directory(new java.io.File("/yourDirectory")).deleteRecursively()
```

File: src/main/scala/microsites/MicrositeKeys.scala (hunk 3 above)
Comment (calvellido) [reply src/main/scala/microsites/MicrositeKeys.scala]: Just curious, what is the benefit?  This is the format being printed now:

```json
[
  {
    "name" : "jv-116-multiversion",
    "own" : true
  },
  {
    "name" : "v0.9.7",
    "own" : false
  }
]
```

I guess `asJson` is already using `spaces2` internally?

https://github.com/circe/circe/blob/master/modules/core/shared/src/main/scala/io/circe/Json.scala#L188

File: src/main/scala/microsites/MicrositeKeys.scala (hunk 2 above)
Comment (calvellido) [reply src/main/scala/microsites/MicrositeKeys.scala]: Addressed in https://github.com/47deg/sbt-microsites/pull/379/commits/22cdc14758ba83507df43b91a8b6189d20f155a1

File: src/main/scala/microsites/MicrositeKeys.scala (hunk 4 above)
Comment (calvellido) [reply src/main/scala/microsites/MicrositeKeys.scala]: Addressed in https://github.com/47deg/sbt-microsites/pull/379/commits/41e996fd7cf2934d9ab7d07d58d722506ab70b9b

File: src/main/scala/microsites/MicrositeKeys.scala (hunk 6 above)
Comment (calvellido) [reply src/main/scala/microsites/MicrositeKeys.scala]: Addressed in https://github.com/47deg/sbt-microsites/pull/379/commits/41e996fd7cf2934d9ab7d07d58d722506ab70b9b

File: src/main/scala/microsites/MicrositeKeys.scala (hunk 3 above)
Comment (juanpedromoreno) [reply src/main/scala/microsites/MicrositeKeys.scala]: Yup, it seems like :) . Thanks for checking.

File: docs/docs/docs/build-the-microsite.md (hunk 7 above)
Comment (juanpedromoreno) [reply docs/docs/docs/build-the-microsite.md]: Makes sense

**Pull request merged**